http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/DiskServiceInterp.scala

```diff
@@ -0,0 +1,233 @@
+  DeletePersistentDisk,
+  ModifyPersistentDisk,
+  ReadPersistentDisk
+}
+import org.broadinstitute.dsde.workbench.leonardo.model.ProjectAction._
+import org.broadinstitute.dsde.workbench.leonardo.monitor.LeoPubsubMessage
+import org.broadinstitute.dsde.workbench.leonardo.monitor.LeoPubsubMessage.{
+  CreateDiskMessage,
+  DeleteDiskMessage,
+  UpdateDiskMessage
+}
+import org.broadinstitute.dsde.workbench.model.google.GoogleProject
+import org.broadinstitute.dsde.workbench.model.{UserInfo, WorkbenchEmail}
+
+import scala.concurrent.ExecutionContext
+
+class DiskServiceInterp[F[_]: Parallel](config: PersistentDiskConfig,
+                                        authProvider: LeoAuthProvider[F],
+                                        serviceAccountProvider: ServiceAccountProvider[F],
+                                        dockerDAO: DockerDAO[F],
+                                        googleStorageService: GoogleStorageService[F],
+                                        publisherQueue: fs2.concurrent.Queue[F, LeoPubsubMessage])(
+  implicit F: Async[F],
+  log: StructuredLogger[F],
+  dbReference: DbReference[F],
+  ec: ExecutionContext
+) extends DiskService[F] {
+
+  override def createDisk(
+    userInfo: UserInfo,
+    googleProject: GoogleProject,
+    diskName: DiskName,
+    req: CreateDiskRequest
+  )(implicit as: ApplicativeAsk[F, AppContext]): F[Unit] =
+    for {
+      ctx <- as.ask
+      hasPermission <- authProvider.hasProjectPermission(userInfo, CreatePersistentDisk, googleProject)
+      _ <- if (hasPermission) F.unit else F.raiseError[Unit](AuthorizationError(Some(userInfo.userEmail)))
+      // Grab the service accounts from serviceAccountProvider for use later
+      clusterServiceAccountOpt <- serviceAccountProvider
+        .getClusterServiceAccount(userInfo, googleProject)
+      petSA <- F.fromEither(
+        clusterServiceAccountOpt.toRight(new Exception(s"user ${userInfo.userEmail.value} doesn't have a PET SA"))
+      )
+
+      diskOpt <- persistentDiskQuery.getActiveByName(googleProject, diskName).transaction
+
+      _ <- diskOpt match {
+        case Some(c) => F.raiseError[Unit](PersistentDiskAlreadyExistsException(googleProject, diskName, c.status))
+        case None =>
+          for {
+            samResourceId <- F.delay(DiskSamResourceId(UUID.randomUUID().toString))
+            disk <- F.fromEither(
+              convertToDisk(userInfo, petSA, googleProject, diskName, samResourceId, config, req, ctx.now)
+            )
+            _ <- authProvider
+              .notifyPersistentDiskCreated(samResourceId, userInfo.userEmail, googleProject)
+              .handleErrorWith { t =>
+                log.error(t)(
+                  s"[${ctx.traceId}] Failed to notify the AuthProvider for creation of persistent disk ${disk.projectNameString}"
+                ) >> F.raiseError(t)
+              }
+
+            _ <- persistentDiskQuery.save(disk).transaction
```

Comment:
We may want to enqueue _before_ persisting the DB record. Reason is we only want to save the record if enqueuing succeeds.

Reply:
We need to get `diskId` by saving the disk before we enqueue it, so I don't think I can flip the order

Reply:
in my other PR, we're introducing a `pre` status to represent the period when the even is not being pushed to pubsub yet

Reply:
Gotcha, I can hold off for now and add a `precreating` status to disks before merging.